mixer/pkg/il/runtime/externs.go

```diff
@@ -111,6 +136,158 @@ func externTimestampEqual(t1 time.Time, t2 time.Time) bool {
+	}
+	return s, err
+}
+
+// This profile converts the string for lookup.
+var externDnsNameEqualProfile = idna.New(idna.MapForLookup(),
+	idna.BidiRule())
+
+func externDnsNameEqual(n1 string, n2 string) (bool, error) {
+	var err error
+
+	if n1, err = externDnsNameEqualProfile.ToUnicode(n1); err != nil {
+		return false, err
+	}
+
+	if n2, err = externDnsNameEqualProfile.ToUnicode(n2); err != nil {
+		return false, err
+	}
+
+	if n1[len(n1)-1] == '.' && n2[len(n2)-1] != '.' {
+		n1 = n1[:len(n1)-1]
+	}
+	if n2[len(n2)-1] == '.' && n1[len(n1)-1] != '.' {
+		n2 = n2[:len(n2)-1]
+	}
+
+	return n1 == n2, nil
+}
+
+func externEmail(in string) (string, error) {
+	a, err := mail.ParseAddress(in)
+	if err != nil {
+		return "", fmt.Errorf("error converting '%s' to e-mail: '%v'", in, err)
+	}
+
+	if a.Name != "" {
+		return "", fmt.Errorf("error converting '%s' to e-mail: display names are not allowed", in)
+	}
+
+	// Also check through the dns name logic to ensure that this will not cause any breaks there, when used for
+	// comparison.
+
+	_, domain := getEmailParts(a.Address)
+
+	_, err = externDnsName(domain)
+	if err != nil {
+		return "", fmt.Errorf("error converting '%s' to e-mail: '%v'", in, err)
+	}
+
+	return in, nil
+}
+
+func externEmailEqual(e1 string, e2 string) (bool, error) {
+	a1, err := mail.ParseAddress(e1)
+	if err != nil {
+		return false, err
+	}
+
+	a2, err := mail.ParseAddress(e2)
+	if err != nil {
+		return false, err
+	}
+
+	local1, domain1 := getEmailParts(a1.Address)
+	local2, domain2 := getEmailParts(a2.Address)
+
+	domainEq, err := externDnsNameEqual(domain1, domain2)
+	if err != nil {
+		return false, fmt.Errorf("error comparing e-mails '%s' and '%s': %v", e1, e2, err)
+	}
+
+	if !domainEq {
+		return false, nil
+	}
+
+	return local1 == local2, nil
+}
+
+func externUri(in string) (string, error) {
+	if in == "" {
+		return "", errors.New("error converting string to url: empty string")
+	}
+	_, err := url.Parse(in)
+	if err != nil {
+		return "", fmt.Errorf("error converting string to url '%s': '%v'", in, err)
+	}
+	return in, err
+}
+
+func externUriEqual(u1 string, u2 string) (bool, error) {
+	url1, err := url.Parse(u1)
+	if err != nil {
+		return false, fmt.Errorf("error converting string to url '%s': '%v'", u1, err)
+	}
+
+	url2, err := url.Parse(u2)
+	if err != nil {
+		return false, fmt.Errorf("error converting string to url '%s': '%v'", u2, err)
+	}
+
+	// Try to apply as much normalization logic as possible.
+	if url1.Scheme != url2.Scheme {
+		return false, nil
+	}
+
+	if url1.Scheme == "http" || url1.Scheme == "https" {
+		// Special case http(s) URLs
+
+		dnsEq, err := externDnsNameEqual(url1.Hostname(), url2.Hostname())
+		if err != nil {
+			return false, err
+		}
+
+		if !dnsEq {
+			return false, nil
+		}
+
+		if url1.Port() != url2.Port() {
+			return false, nil
+		}
+
+		// normalize url2 to url1's hostname.
+		url1.Host = url2.Host
+	}
+
+	return url1.String() == url2.String(), nil
```

Comment:
Case insensitive?

Reply:
The case sensitive parts are (for http/https) are normalized within the if block. Other parts are case sensitive.